python/ray/tune/variant_generator.py

```diff
@@ -0,0 +1,275 @@
+            else:
+                argv.append(json.dumps(v))
+        return argv
+
+    parser = make_parser()
+    i = 0
+    for _ in range(unresolved_spec.get("repeat", 1)):
+        for resolved_vars, spec in generate_variants(unresolved_spec):
+            args = parser.parse_args(to_argv(spec))
+            if resolved_vars:
+                experiment_tag = "{}_{}".format(i, resolved_vars)
+            else:
+                experiment_tag = str(i)
+            i += 1
+            yield Trial(
+                env_creator=spec.get("env", lambda: None),
+                alg=spec.get("alg", "script"),
+                config=spec.get("config", {}),
+                local_dir=os.path.join(args.local_dir, output_path),
+                experiment_tag=experiment_tag,
+                resources=json_to_resources(spec.get("resources", {})),
+                stopping_criterion=spec.get("stop", {}),
+                checkpoint_freq=args.checkpoint_freq,
+                restore_path=spec.get("restore"),
+                upload_dir=args.upload_dir)
+
+
+def generate_variants(unresolved_spec):
+    """Generates variants from a spec (dict) with unresolved values.
+
+    There are two types of unresolved values:
+
+        Grid search: These define a grid search over values. For example, the
+        following grid search values in a spec will produce six distinct
+        variants in combination:
+
+            "activation": grid_search(["relu", "tanh"])
+            "learning_rate": grid_search([1e-3, 1e-4, 1e-5])
+
+        Lambda functions: These are evaluated to produce a concrete value, and
+        can express dependencies or conditional distributions between values.
+        They can also be used to express random search (e.g., by calling
+        into the `random` or `np` module).
+
+            "cpu": lambda spec: spec.config.num_workers
+            "batch_size": lambda spec: random.uniform(1, 1000)
+
+        It is also possible to nest the two, e.g. have a lambda function
+        return a grid search or vice versa, as long as there are no cyclic
+        dependencies between unresolved values.
+
+    Finally, to support defining specs in plain JSON / YAML, grid search
+    and lambda functions can also be defined alternatively as follows:
+
+        "activation": {"grid_search": ["relu", "tanh"]}
+        "cpu": {"eval": "spec.config.num_workers"}
+    """
+    for resolved_vars, spec in _generate_variants(unresolved_spec):
+        assert not _unresolved_values(spec)
+        yield _format_vars(resolved_vars), spec
+
+
+def grid_search(values):
+    """Convenience method for specifying grid search over a value."""
+
+    return {"grid_search": values}
+
+
+_STANDARD_IMPORTS = {
+    "random": random,
+    "np": numpy,
+}
+
+_MAX_RESOLUTION_PASSES = 20
+
+
+def _format_vars(resolved_vars):
+    out = []
+    for path, value in sorted(resolved_vars.items()):
+        if path[0] in ["alg", "env", "resources"]:
```

Comment:
would be nice to give warning for this or else if someone puts "alg" in their model config it would fail silently and no idea what's going on